### PR TITLE
Independent Component Analysis (ICA)

### PR DESCRIPTION
**Pull request by VasanthakumarV**
This PR is for the port of sklearn's [FastICA](https://github.com/scikit-learn/scikit-learn/blob/0fb307bf3/sklearn/decomposition/_fastica.py#L304).

### CODE REVIEW

File: linfa-reduction/src/fast_ica/mod.rs [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+        }
+
+        Ok(w)
+    }
+
+    // Symmetric decorrelation
+    //
+    // W <- (W * W.T)^{-1/2} * W
+    fn sym_decorrelation<A: Float>(w: &Array2<A>) -> Array2<A> {
+        let (eig_val, eig_vec) = w.dot(&w.t()).eigh(UPLO::Upper).unwrap();
```
Comment (bytesnake) [inline linfa-reduction/src/fast_ica/mod.rs:185]: the eigenvalue decomposition could theoretical fail

File: linfa-reduction/src/fast_ica/mod.rs [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+        let eig_val = eig_val.mapv(|x| A::from(x).unwrap());
+
+        let tmp = &eig_vec
+            * &(eig_val
+                .mapv(num_traits::Float::sqrt)
+                .mapv(num_traits::Float::recip))
```
Comment (bytesnake) [inline linfa-reduction/src/fast_ica/mod.rs:191]: dont' invert without clipping low singular values close to zero, you could for example clip everything below `1e-7`

File: linfa-reduction/src/fast_ica/mod.rs [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+#[derive(Debug)]
+pub struct FittedFastIca<A> {
+    mean: Array1<A>,
+    components: Array2<A>,
+}
+
+impl<A: Float> FittedFastIca<A> {
+    /// Recover the sources
+    pub fn transform(&self, x: &Array2<A>) -> Array2<A> {
+        let xcentered = x - &self.mean.to_owned().insert_axis(Axis(0));
```
Comment (bytesnake) [inline linfa-reduction/src/fast_ica/mod.rs:208]: same as above, use `view().insert_axis(Axis(0))`

File: linfa-reduction/src/fast_ica/mod.rs [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+    fn ica_parallel<A: Float>(&self, x: &Array2<A>, w: &Array2<A>) -> Result<Array2<A>, String> {
+        let mut w = Self::sym_decorrelation(&w);
+
+        let p = x.ncols() as f64;
+
+        for _ in 0..self.max_iter {
+            let (gwtx, g_wtx) = self.gfunc.exec(&w.dot(x))?;
+
+            let lhs = gwtx.dot(&x.t()).mapv(|x| x / A::from(p).unwrap());
+            let rhs = &w * &g_wtx.insert_axis(Axis(1));
```
Comment (bytesnake) [inline linfa-reduction/src/fast_ica/mod.rs:159]: isn't a `1/N` missing here, because you want to calculate the mean value of the gradient?

File: linfa-reduction/src/fast_ica/mod.rs [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+            let wnew = Self::sym_decorrelation(&(lhs - rhs));
+
+            let lim = *wnew
+                .dot(&w.t())
+                .diag()
```
Comment (bytesnake) [inline linfa-reduction/src/fast_ica/mod.rs:164]: you want to calculate dot product between every entry `w` here right? You could use an iterator
```rust
wnew.outer_iter().zip(w.outer_iter()).map(|(a, b)| a.dot(&b)).collect::<Array1<_>>()
```
this has only complexity O(n^2)

File: linfa-reduction/src/fast_ica/mod.rs [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+        let p = x.ncols() as f64;
+
+        for _ in 0..self.max_iter {
+            let (gwtx, g_wtx) = self.gfunc.exec(&w.dot(x))?;
+
+            let lhs = gwtx.dot(&x.t()).mapv(|x| x / A::from(p).unwrap());
+            let rhs = &w * &g_wtx.insert_axis(Axis(1));
+            let wnew = Self::sym_decorrelation(&(lhs - rhs));
+
+            let lim = *wnew
```
Comment (bytesnake) [inline linfa-reduction/src/fast_ica/mod.rs:162]: I would a sentence here like `calculate maximal weight vector change with dot product`, because it isn't obvious right away

File: linfa-reduction/src/fast_ica/mod.rs [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+                        test_fast_ica($gfunc);
+                    }
+                )*
+            }
+        }
+    }
+
+    // Tests to make sure all of the `GFunc`'s non-linear functions and the
+    // model itself performs well
+    fast_ica_tests! {
```
Comment (bytesnake) [inline linfa-reduction/src/fast_ica/mod.rs:306]: nice :+1:

File: linfa-reduction/src/fast_ica/mod.rs [hunk 8]
```diff
@@ -0,0 +1,382 @@
+//! Fast algorithm for Independent Component Analysis (ICA)
+//!
+//! ICA separates mutivariate signals into their additive, indipendent subcomponents.
```
Comment (bytesnake) [inline linfa-reduction/src/fast_ica/mod.rs:3]: indipendent -> independent

**Review by bytesnake — COMMENTED** [review]
nice work :+1: you could reduce the matrix multiplication complexity in the update step, but everything else is looking good. Would move to own crate though

File: linfa-reduction/README.md [hunk 9]
```diff
@@ -11,6 +11,7 @@
 `linfa-reduction` currently provides an implementation of the following dimensional reduction methods: 
 - Diffusion Mapping
 - Principal Component Analysis (PCA)
+- Independent Component Analysis (ICA)
```
[the remark this answers is not in the document]
Comment (VasanthakumarV) [reply linfa-reduction/README.md]: I have moved everything to a new crate `linfa-ica`

File: linfa-reduction/examples/fast_ica.rs [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    // Create sample dataset for the model
+    // `s_original` has the unmixed sources (we merely have it to save to disk)
+    // `s` is the mixed source that will be unmixed using ICA
+    // Shape of the data will be (2000 x 2)
+    let (sources_original, sources_mixed) = create_data();
+
+    // Fit the model
+    // `ncomponents` is not set, it will be automatically be assigned 2 from
+    // the input
+    let ica = FastIca::new().set_gfunc(GFunc::Logcosh(1.0));
```
[the remark this answers is not in the document]
Comment (VasanthakumarV) [reply linfa-reduction/examples/fast_ica.rs]: Done

File: linfa-reduction/src/fast_ica/mod.rs [hunk 11]
```diff
@@ -0,0 +1,382 @@
...
+            ncomponents: None,
+            gfunc: GFunc::Logcosh(1.),
+            max_iter: 200,
+            tol: 1e-4,
+            random_state: None,
+        }
+    }
+
+    /// Set the number of components to use, if not set all are used
+    pub fn set_ncomponents(mut self, ncomponents: usize) -> Self {
```
[the remark this answers is not in the document]
Comment (VasanthakumarV) [reply linfa-reduction/src/fast_ica/mod.rs]: Done

File: linfa-reduction/src/fast_ica/mod.rs [hunk 12]
```diff
@@ -0,0 +1,382 @@
...
+        if ncomponents > nsamples.min(nfeatures) {
+            return Err(
+                "ncomponents cannot be greater than the min(sample size, features size)"
+                    .to_string(),
+            );
+        }
+
+        // We center the input by subtracting the mean of its features
+        let xmean = x.mean_axis(Axis(0)).unwrap();
+        let mut xcentered = x - &xmean.to_owned().insert_axis(Axis(0));
```
[the remark this answers is not in the document]
Comment (VasanthakumarV) [reply linfa-reduction/src/fast_ica/mod.rs]: Done

File: linfa-reduction/src/fast_ica/mod.rs [hunk 13]
```diff
@@ -0,0 +1,382 @@
...
+        // the components
+        let k = match xcentered.svd(true, false).unwrap() {
+            (Some(u), s, _) => {
+                let s = s.mapv(|x| A::from(x).unwrap());
+                (u.slice(s![.., ..nsamples.min(nfeatures)]).to_owned() / s)
+                    .t()
+                    .slice(s![..ncomponents, ..])
+                    .to_owned()
+            }
+            _ => unreachable!(),
```
[the remark this answers is not in the document]
Comment (VasanthakumarV) [reply linfa-reduction/src/fast_ica/mod.rs]: Done, have created a new error type for this case

File: linfa-reduction/src/fast_ica/mod.rs (hunk 1 above)
Comment (VasanthakumarV) [reply linfa-reduction/src/fast_ica/mod.rs]: Now propagating the error instead of unwrapping

File: linfa-reduction/src/fast_ica/mod.rs (hunk 2 above)
Comment (VasanthakumarV) [reply linfa-reduction/src/fast_ica/mod.rs]: Done

File: linfa-reduction/src/fast_ica/mod.rs (hunk 3 above)
Comment (VasanthakumarV) [reply linfa-reduction/src/fast_ica/mod.rs]: Done

File: linfa-reduction/src/fast_ica/mod.rs (hunk 4 above)
Comment (VasanthakumarV) [reply linfa-reduction/src/fast_ica/mod.rs]: `p` here is N, x is transposed so `p` which is `x.ncols()` would be N

File: linfa-reduction/src/fast_ica/mod.rs (hunk 5 above)
Comment (VasanthakumarV) [reply linfa-reduction/src/fast_ica/mod.rs]: Done, thanks

File: linfa-reduction/src/fast_ica/mod.rs (hunk 6 above)
Comment (VasanthakumarV) [reply linfa-reduction/src/fast_ica/mod.rs]: I have added a comment about `lim` calculation if that is what you are referring

File: linfa-reduction/src/fast_ica/mod.rs (hunk 8 above)
Comment (VasanthakumarV) [reply linfa-reduction/src/fast_ica/mod.rs]: Corrected

**Comment by bytesnake** [comment]
lets merge this then :tada: the failure is unrelated to fastica, will fix this in a separate commit
